Lib/functools.py

```diff
@@ -963,13 +964,27 @@ def __isabstractmethod__(self):
 
 
 class cached_property:
-    def __init__(self, func):
+    def __init__(self, func=None, *, lock=True):
         self.func = func
         self.attrname = None
         self.__doc__ = func.__doc__
-        self.lock = RLock()
+        if lock:
+            warn("Locking cached_property is deprecated; use @cached_property(lock=False)",
+                 PendingDeprecationWarning, 2)
```

Comment:
Maybe use [`warnings._deprecated`](https://discuss.python.org/t/introducing-warnings-deprecated/14856)?

Reply:
I don't think `warnings._deprecated` works here, because it assumes the thing that is deprecated is a single method/function/class with a name. What would you pass in for the name argument of `warnings._deprecated` in this case?

Reply:
You could do this:

```diff
diff --git a/Lib/functools.py b/Lib/functools.py
index 8ad150d853..5eb772e6b4 100644
--- a/Lib/functools.py
+++ b/Lib/functools.py
@@ -969,9 +969,12 @@ def __init__(self, func=None, *, lock=True):
         if func is not None:
             self.__doc__ = func.__doc__
         if lock:
-            from warnings import warn
-            warn("Locking cached_property is deprecated; use @cached_property(lock=False)",
-                 PendingDeprecationWarning, 2)
+            import warnings
+            warnings._deprecated(
+                "lock=True default behaviour of cached_property",
+                "Locking cached_property is deprecated; use @cached_property(lock=False)",
+                remove=(3, 14)
+            )
         self.lock = RLock() if lock else None
```

And then you'd get this behaviour:

```pycon
C:\Users\alexw\coding\cpython>python
Running Debug|x64 interpreter...
Python 3.12.0a0 (main, Oct  9 2022, 08:02:35) [MSC v.1932 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>>
>>> # Behaviour on 3.12:
>>>
>>> import functools
>>> x = functools.cached_property(lambda self: 42)
<stdin>:1: DeprecationWarning: Locking cached_property is deprecated; use @cached_property(lock=False)
>>>
>>> # Behaviour when we get to 3.14 beta if the default hasn't been changed yet:
>>>
>>> import sys
>>> sys.version_info = 3, 14, 0, "beta"
>>> del sys.modules["warnings"]; del sys.modules["functools"]
>>> import functools
>>> x = functools.cached_property(lambda self: 42)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\alexw\coding\cpython\Lib\functools.py", line 973, in __init__
    warnings._deprecated(
  File "C:\Users\alexw\coding\cpython\Lib\warnings.py", line 511, in _deprecated
    raise RuntimeError(msg)
RuntimeError: 'lock=True default behaviour of cached_property' was slated for removal after Python 3.14 alpha
```

Reply:
Also, any reason you're using `PendingDeprecationWarning` instead of `DeprecationWarning`, btw?

Reply:
> You could do this

Ok, agreed, using a full noun phrase as the name does seem to work fine. I can make this change.

> any reason you're using PendingDeprecationWarning instead of DeprecationWarning

I thought the general preference was to provide one release of `PendingDeprecationWarning` before moving to `DeprecationWarning` in the next release, and then removing in the release after that?